### PR TITLE
support phpstan 1.0; drop support for phpstan < 1.0

### LINKED ISSUE #25: Prepare for PHPStan 1.0
Hello everyone 👋 

[I announced today](https://twitter.com/OndrejMirtes/status/1448544146944401409) that PHPStan 1.0 is going to be released on November 1st 2021.

I'm approaching you as one of the most popular PHPStan extensions. I'd love if you could prepare your code for PHPStan 1.0 in advance so that it's ready to release on the same day.

Here's a brief guide how to approach the upgrade:

1) Create a branch 🌴 
2) Update your composer.json to `"phpstan/phpstan": "^1.0"`, add `"minimum-stability": "dev"` and `"prefer-stable": true` if necessary.
3) Update your code with the BC breaks below in mind. 🔧 
4) Fix the code so that it passes PHPStan's analysis 🤓 
5) Wait for PHPStan 1.0 release on November 1st, merge your branch and tag the next major version 👍 

Thank you!

-----------

Here are the BC breaks. The list is huge but most of those have very little impact.

There are new rules around using PHPStan internals in regard to backward compatibility promise: https://phpstan.org/developing-extensions/backward-compatibility-promise

It's possible that not everything you use is covered by it - so I'm here to help you to transition to correct usage, or add some `@api` annotations in PHPStan itself so that more is covered by the promise. Let me know!

BC breaks for end-users
=======================

* Level `max` is now level 9. Feel free to update `max` to `8` if your code isn't ready.
* Description of some types was updated to match the syntax in PHPDocs. This includes `numeric-string` (it was previously described as `string&numeric`), and also [array shapes](https://phpstan.org/writing-php-code/phpdoc-types#array-shapes).
* Removed `baselineNeon` error formatter, use `--generate-baseline` CLI option instead (https://github.com/phpstan/phpstan-src/commit/492cfbcd74b540eda1520e11b67d7be00cbc7c31)
* Removed `polluteCatchScopeWithTryAssignments` config parameter (https://github.com/phpstan/phpstan-src/commit/8933c7e28c2ebb1fd69584c031f579cd53ddf4ae)
* Removed deprecated `autoload_files` parameter - use [Discovering Symbols](https://phpstan.org/user-guide/discovering-symbols) instead (https://github.com/phpstan/phpstan-src/commit/7a21246cae9dd7968bf7bef92223b53f5d681b72)
* Removed deprecated `autoload_directories` parameter - use [Discovering Symbols](https://phpstan.org/user-guide/discovering-symbols) instead (https://github.com/phpstan/phpstan-src/commit/f67b48a71220b295a1704fe9749492e5de9ee18f)
* Removed `bootstrap` parameter - use `bootstrapFiles` instead (https://github.com/phpstan/phpstan-src/commit/1baa29425d2e170af8ac53e4e579c34c25b640e4)
* Moved `implicitThrows` configuration parameter to `exceptions.implicitThrows` (https://github.com/phpstan/phpstan-src/commit/96b7c48025dcbd5a709cea71ee4e9e73770e0d54)
* Removed `--paths-file` CLI option (https://github.com/phpstan/phpstan-src/commit/5670cf221723eb652643623ad487351cb187f6c3)
* Removed `dump-deps` command (https://github.com/phpstan/phpstan-src/commit/9c7017c697c29668f8a7c766236e597737f99f91)
* Deprecated `excludes_analyse` option, use `excludePaths` instead (https://github.com/phpstan/phpstan-src/commit/d25c5e5a6c62a4537d4a108707ff26b5bea687d9)

The following are interesting only if you create a custom ruleset in your configuration file:

* Removed `DeadCatchRule`, replaced by `CatchWithUnthrownExceptionRule` (https://github.com/phpstan/phpstan-src/commit/4dba60bacbf3ac0efe8627fa32177ee21cdd3eca)
* Removed `VariableCertaintyInIssetRule`, replaced by `IssetRule` (https://github.com/phpstan/phpstan-src/commit/2e858dee389d4843e4d1a7dc1bc565c264be247f)
* Removed `MissingClosureNativeReturnTypehintRule`, no longer needed thanks to type inference (https://github.com/phpstan/phpstan-src/commit/1c34d8dc855cb5f4cf5c8b033b7155eeed0e277b)
* Rename rules with typos (https://github.com/phpstan/phpstan-src/commit/003ab1aea558085cd27636a0b95a4c9f0e062e9a)

BC breaks for extension developers
=======================

* Extensions are checked with rules for correct usage of PHPStan internals according to [backward compatibility promise](https://phpstan.org/developing-extensions/backward-compatibility-promise).
* Removed `CompoundTypeHelper` (https://github.com/phpstan/phpstan-src/commit/145c4e3af4045b074b1d6e697022a835433cd5f2)
* Removed `CommentHelper` (https://github.com/phpstan/phpstan-src/commit/ebad6f61b9a65b68c200f7ac3a444ce198149396)
* Removed `DependencyDumper` (https://github.com/phpstan/phpstan-src/commit/9c7017c697c29668f8a7c766236e597737f99f91)
* Removed `PHPStan\Reflection\Generic\ResolvedFunctionVariant`, replaced by `PHPStan\Reflection\ResolvedFunctionVariant` (https://github.com/phpstan/phpstan-src/commit/1cc6c8175a6d92cf717cd553506c9da8a2dd72cc)
* Removed `ClassReflection::getNativeMethods()`, use `getNativeReflection()` instead (https://github.com/phpstan/phpstan-src/commit/d2c1446b3b0f581502019c522ccb521b37e7dc15)
* Removed `PhpPropertyReflection::hasPhpDoc()`, replaced by `hasPhpDocType()` (https://github.com/phpstan/phpstan-src/commit/bedd5be99107254f42efb7728c9ce3a7b958e33e)
* `NodeDependencies` no longer iterable (https://github.com/phpstan/phpstan-src/commit/f76875a84d8dcca8f9d67c08b6492c609e9aff98)
* Renamed `TestCase` to `PHPStanTestCase` ([#634](https://github.com/phpstan/phpstan-src/pull/634)), thanks @frankdejonge!
* `StaticType` and `ThisType` - require `ClassReflection` in constructor (https://github.com/phpstan/phpstan-src/commit/7aabc848351689f18a1db99598df8d5b4944aa87)
* `PHPStanTestCase` - extensions can no longer be provided by overriding methods from `PHPStanTestCase`. Use `getAdditionalConfigFiles()` instead. (https://github.com/phpstan/phpstan-src/commit/65efd93f764e9db076851ef262179fe670f1ab38, https://github.com/phpstan/phpstan-src/commit/239291a91089da6684d8c52e58b9a77ceb178071)
* Removed some unused internal helper methods from Broker (https://github.com/phpstan/phpstan-src/commit/d57815103091d8f25347b8441da3ed099f70d7af)
* Changed return types of various methods that had `|false` to `|null` (https://github.com/phpstan/phpstan-src/commit/629ccf6b02146e1cbf0499982358c595e3d6edd5)
* Type generalization precision is now a required argument (https://github.com/phpstan/phpstan-src/commit/65681033ff625c67a478212a0cd6da29cd5ab5fa)
* Some constructor parameters are now required (https://github.com/phpstan/phpstan-src/commit/1f4062fe2d0d5406e9b8a782f99b6da61eceafde)
* Deprecated `BrokerAwareExtension` (https://github.com/phpstan/phpstan-src/commit/db2f7fb9e7928de68665ca09a3e1002d8a329b33)
* Deprecated `PHPStanTestCase::createBroker()`. Use `createReflectionProvider()` instead. (https://github.com/phpstan/phpstan-src/commit/1e5cf58e07446e634803c3be87e7c41fdd903c7c)
* Deprecated `PHPStan\Broker\Broker`. Use `PHPStan\Reflection\ReflectionProvider` instead. (https://github.com/phpstan/phpstan-src/commit/c7755948c124cfbcb1220ab912a6eaa863cdf5fb)
* Deprecated `PHPStan\Broker\Broker::getInstance()`. Use `PHPStan\Reflection\ReflectionProviderStaticAccessor` instead. (https://github.com/phpstan/phpstan-src/commit/4e7d60d74b3fda7dc4d3e0c0798b7fd55a7b32f2)

### PR DESCRIPTION
**Pull request by staabm**
[phpstan 0.x will no longer be supported](https://twitter.com/OndrejMirtes/status/1455083738564542467). prepare the rule for phpstan 1.0+

Fixes #25

### CODE REVIEW

**Comment by staabm** [comment]
locally the tests pass now. I have 0 experience with this lib, please take it with a grain of salt

File: tests/Rules/CallMethodRuleTest.php [hunk 1]
```diff
@@ -14,10 +17,10 @@ class CallMethodRuleTest extends RuleTestCase
     protected function getRule(): Rule
     {
         $broker = $this->createBroker();
-        $ruleLevelHelper = new RuleLevelHelper($broker, true, true, true);
+        $ruleLevelHelper = new RuleLevelHelper($broker, true, true, true, false);
         return new CallMethodsRule(
             $broker,
-            new FunctionCallParametersCheck($ruleLevelHelper, true, true, true, true),
+            new FunctionCallParametersCheck($ruleLevelHelper, new NullsafeCheck(), new PhpVersion(PHP_VERSION_ID), new UnresolvableTypeHelper(), true, false, false, false),
```
Comment (staabm) [inline tests/Rules/CallMethodRuleTest.php:23]: especially I am not sure about these 3 new added `false` passed for the 3 new args:
grafik

Comment (p4veI) [reply tests/Rules/CallMethodRuleTest.php:23]: I'm just looking around, but from looking into phpstan core, I don't see that the booleans are added to the FunctionCallParametersCheck, only the last boolean was removed.

Frankly, it seems odd that the instantiation of FunctionCallParametersCheck did not include the nullsafe check, phpversion and helper before and was just passing the booleans instead..

Feels like it should either stay the same as:
`...true, true, true, true` unless there's a reason for disabling the checks..

Anyways, I temporarily updated phpstan-safe-rule in my project with this fork and it seems it works just as well as previously. Hope this helps..

**Comment by oneserv-heuser** [comment]
Closes https://github.com/ergebnis/phpstan-rules/issues/379

**Comment by oneserv-heuser** [comment]
Fixes #25

**Comment by mfn** [comment]
I can confirm it "just works" for my use-case; I referenced the last commit in this PR:
```js
    "thecodingmachine/phpstan-safe-rule": "dev-master#a630c7bf9f3060b6e3dc74a4d9ef5edc0feec052",
```
and added this repository to my `composer.json`:
```js
    {
      "type": "package",
      "package": {
        "name": "thecodingmachine/phpstan-safe-rule",
        "type": "library",
        "version": "dev-master",
        "dist": {
          "url": "https://codeload.github.com/thecodingmachine/phpstan-safe-rule/legacy.zip/a630c7bf9f3060b6e3dc74a4d9ef5edc0feec052",
          "type": "zip"
        },
        "autoload": {
          "psr-4": {
            "TheCodingMachine\\Safe\\PHPStan\\": "src/"
          }
        }
      }
    }

```

**Comment by OskarStark** [comment]
Can someone please merge and release this?

Thanks

**Comment by moufmouf** [comment]
Hey @staabm ,

Thanks a lot for the PR!

Could you also please update the `composer.json` file to switch the alias master from "1.0.x" to "1.1.x"?

https://github.com/thecodingmachine/phpstan-safe-rule/blob/master/composer.json#L38-L40

This way, the "1.0.x" tags will stay associated with PHPStan <= 0.12 and the 1.1.x tags will remain associated with PHPStan 1

**Comment by staabm** [comment]
@moufmouf done

**Comment by moufmouf** [comment]
So quick! Thanks a lot for the PR :heart:

**Comment by moufmouf** [comment]
And it's tagged: https://github.com/thecodingmachine/phpstan-safe-rule/releases/tag/v1.1.0

**Comment by OskarStark** [comment]
Bär, thank you very much! 🎉

**Pull request merged**